### PR TITLE
feat: Add a metric ingestion time SM sanitization

### PR DESCRIPTION
**Pull request by cstyan**
Adding a per-tenant metric for when a tenant has structured metadata sanitized at ingestion time. This would be less spammy than a log line, and while the `tenant` label could be high cardinality it's unlikely to be as the % of users using SM and sending SM that has invalid characters is likely low.

Some of the discussion around https://github.com/grafana/loki/pull/15113 has been that the query time sanitization of the SM name and value (value would be sanitized in that PR, name sanitization is already in place) is that query time sanitization is expensive, and we'd like to move that functionality to be on a per-tenant basis.

The goal here is to inform us which users have been sending invalid SM so that we modify the query time SM sanitization code blocks so that they're only run on a per-tenant basis based on per-tenant overrides. With this metric in place we would be able to have a good idea ahead of time of which users have been sending invalid SM before we make the swap over so that we can set the per-tenant config in advance of the rollout.

### CODE REVIEW

File: pkg/distributor/distributor.go [hunk 1]
```diff
@@ -527,9 +533,14 @@ func (d *Distributor) Push(ctx context.Context, req *logproto.PushRequest) (*log
...
 				}
 
+				var normalized string
 				structuredMetadata := logproto.FromLabelAdaptersToLabels(entry.StructuredMetadata)
 				for i := range entry.StructuredMetadata {
-					structuredMetadata[i].Name = otlptranslate.NormalizeLabel(structuredMetadata[i].Name)
+					normalized = otlptranslate.NormalizeLabel(structuredMetadata[i].Name)
+					if normalized != structuredMetadata[i].Name {
+						structuredMetadata[i].Name = normalized
+						d.tenantPushSanitizedStructuredMetadata.WithLabelValues(tenantID).Inc()
```
Comment (ashwanthgoli) [inline pkg/distributor/distributor.go:542]: should we also add this for value sanitisation?

File: pkg/distributor/distributor_test.go [hunk 2]
```diff
@@ -1988,5 +1994,7 @@ func TestDistributor_StructuredMetadataSanitization(t *testing.T) {
 		response, err := distributors[0].Push(ctx, &request)
 		require.NoError(t, err)
 		assert.Equal(t, tc.expectedResponse, response)
+		//fmt.Println("counter: ", distributors[0].tenantPushSanitizedStructuredMetadata.WithLabelValues())
```
Comment (ashwanthgoli) [inline pkg/distributor/distributor_test.go]:
```suggestion
```

File: pkg/distributor/distributor.go (hunk 1 above)
Comment (cstyan) [reply pkg/distributor/distributor.go:542]: yep, this just got missed in my `git add --patch` and then I saw the test failing as you posted your comment, pushed the value inc() call

**Review by ashwanthgoli — APPROVED** [review]
lgtm!

**Review by cyriltovena — APPROVED** [review]
LGTM

Not sure we need a metrics for that but why not.

**Pull request merged**